File: Swift/Structures/OrderedList.swift
final class Node{
    var data:Int = 0;
    var next:Node? = nil;
    var prev:Node? = nil;

    init(m_data:Int) {data = m_data;}
}

class OrderedList{
    private var front:Node? = nil;
    private var back:Node? = nil;
    private var m_size:Int = 0;

    func add(data:Int){
        var node:Node = Node(data);
        if isEmpty(){
            self.front = node;
            self.back = node;
        }
        else if data >= self.front!.data {
            node?.next = self.front;
            self.front?.prev = node;
            self.front = node;
        }
        else if data <= self.back!.data {
            self.back?.next = node;
            node?.prev = self.back;
            self.back = node;
        }
        else{
            var ptr:Node? = self.front;

            while(ptr?.next != nil){
                if ptr!.next.data < data{
                    break;
                }
                ptr = ptr?.next;
            }

            if(ptr?.next != nil){
                ptr?.next.prev = node;
                node?.next = ptr?.next;
                node?.prev = ptr;
                ptr?.next = node;
            }
        }
        self.m_size += 1;
    }

    func remove_front(){ // how GC is managed in swift?
        if isEmpty(){
            return;
        }
        self.front = self.front?.next;
        self.m_size -= 1;
    }

    func remove_back(){
        if isEmpty(){
            return;
        }
        self.back = self.back?.prev;
        self.m_size -= 1;
    }

    func remove(data:Int){
        if isEmpty(){
            return;
        }
        else{
            var ptr:Node? = self.front;
            var prev:Node? = self.ptr;

            while (ptr?.next != nil){
                if(ptr!.data == data){
                    prev?.next = ptr?.next;
                    ptr?.next.prev = prev;

                    self.m_size -= 1;
                    return; 
                }
                prev = ptr;
                ptr = ptr?.next;
            }
        }
    }

    func display(){
        var ptr:Node? = self.front;
        while(ptr?.next != nil){
            print(" \(ptr!.data)");
            ptr = ptr?.next;
        }
    }

    func isEmpty()->Bool{return self.front == nil;}

    func size()->Int{return self.m_size; }
}

File: Python/Structures/OrderedList.py
class Node:
    next: Node = None
    prev: Node = None
    data: int

    def __init__(self, data=0) -> None:
        self.data = data


class OrderedList:
    front: Node = None
    back: Node = None
    size: int = 0

    def add(self, data: int) -> None:
        node: Node = None(data)
        if self.is_empty():
            self.front = node
            self.back = node
        elif data > self.front.data:
            self._add_front(node)
        elif data > self.back.data:
            self._add_back(node)
        else:
            ptr: Node = self.front
            while ptr.next is not None:
                if ptr.data > data:
                    break
                ptr = ptr.next

            if ptr.next is not None:
                ptr.next.prev = node
                node.next = ptr.next
                node.prev = ptr
                ptr.next = ptr

        size += 1

    def remove(self, data: int) -> None:
        if self.is_empty():
            raise Exception("Error: element not found in the list.")

        elif self.front.data == data:
            self._remove_front()
        elif self.back.data == data:
            self._remove_back()
        else:
            ptr: Node = self.front
            prev: Node = ptr

            while ptr.next is not None:
                if ptr.data == data:
                    ptr.next = ptr.next
                    ptr.prev = prev

                    ptr = None
                    size -= 1
                    return
                prev = ptr
                ptr = ptr.next

    def display(self) -> None:
        ptr: Node = self.front
        while ptr.next is not None:
            print(ptr.data)
            ptr = ptr.next

    def _add_front(self, node: Node) -> None:
        self.front.prev = node
        node.next = self.front
        self.front = node

    def _add_back(self, node: Node) -> None:
        self.back.next = node
        node.prev = self.back
        back = node

    def _remove_front(self) -> None:
        temp: Node = self.front
        self.front = self.front.next
        temp = None
        size -= 1

    def _remove_back(self) -> None:
        temp: Node = self.back
        self.back = self.back.prev
        temp = None
        size -= 1

    def is_empty(self) -> bool:
        return self.front is None and self.back is None
